gobblin-tunnel/src/test/java/gobblin/tunnel/TestTunnelWithArbitraryTCPTraffic.java

```diff
@@ -0,0 +1,859 @@
+        hashOfMsgsRecvdAtClients.put(client, Hex.encodeHexString(digestMsgsRecvdAtClients.get(client).digest()));
+      }
+
+      System.out.println("\tComparing client sent to server received");
+      assertEquals(hashOfMsgsSentByClients, hashOfMsgsRecvdAtServer);
+
+      System.out.println("\tComparing server sent to client received");
+      for (String hashOfMsgsRecvdAtClient : hashOfMsgsRecvdAtClients.values()) {
+        assertEquals(hashOfMsgsSentByServer, hashOfMsgsRecvdAtClient);
+      }
+      System.out.println("\tDone");
+    }
+  }
+
+  private MockServer startTalkPastServer(final int nMsgs, final Map<String, MessageDigest> digestMsgsRecvdAtServer) throws IOException {
+    return new MockServer() {
+      @Override
+      void handleClientSocket(Socket clientSocket) throws IOException {
+        System.out.println("Writing to client");
+        try {
+          final BufferedOutputStream serverOut = new BufferedOutputStream(clientSocket.getOutputStream());
+          EasyThread clientWriterThread = new EasyThread() {
+            @Override
+            void runQuietly() throws Exception {
+              long t = System.currentTimeMillis();
+              try {
+                for (int i = 0; i < nMsgs; i++) {
+                  serverOut.write(generateMsgFromServer(i).getBytes());
+                  sleepQuietly(2);
+                }
+              } catch (IOException e) {
+                e.printStackTrace();
+              }
+              System.out.println("Server done writing in " + (System.currentTimeMillis() - t) + " ms");
+            }
+          }.startThread();
+          _threads.add(clientWriterThread);
+
+          BufferedReader serverIn = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
+          String line = serverIn.readLine();
+          while (line != null && !line.equals("Goodbye")) {
+            //System.out.println("Client said [" + line.substring(0, 32) + "... ]");
+            String [] tokens = line.split(":", 2);
+            String client = tokens[0];
+            digestMsgsRecvdAtServer.get(client).update(line.getBytes());
+            digestMsgsRecvdAtServer.get(client).update("\n".getBytes());
+            line = serverIn.readLine();
+          }
+          System.out.println("Server done reading");
+          try {
+            clientWriterThread.join();
+          } catch (InterruptedException e) {
+          }
+          serverOut.write("Goodbye\n".getBytes());
+          serverOut.flush();
+          clientSocket.close();
+        } catch (IOException e) {
+          e.printStackTrace();
+          throw e;
+        }
+      }
+    }.start();
+  }
+
+  private static String generateMsgFromServer(int i) {
+    return i + " " + StringUtils.repeat("Babble babble ", 10000) + "\n";
+  }
+
+  // Baseline tests to ensure simultaneous data transfer protocol is fine (disabled for now)
+  @Test(enabled = false, timeOut = 30000)
```

Comment:
It's milliseconds, so this is about 30 seconds. This test is usually done within a second on my box, but I wasn't sure how long it would take on more resource-constrained environments like Travis. Can reduce it to 5000 like other tests if this isn't what is causing the processes to be killed.